File: clients/js/src/features/config_embed_wrap_package.ts
import { Input } from "../input";

import { PolywrapClientConfigBuilder, PolywrapClient } from "@polywrap/client-js";
import { WasmPackage } from "@polywrap/wasm-js";

import path from "path";
import fs from "fs";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    directory: unknown;
    method: unknown;
    args: unknown;
  }>(input);

  const wrapDir = Input.expectRootDir(
    inputObj.directory,
    path.join(__dirname, "../../../../")
  );

  const method = Input.expectString(inputObj.method);
  const args = Input.expectObject<Record<string, unknown>>(
    inputObj.args
  );

  console.log(`Reading wrap.info & wrap.wasm from ${inputObj.directory}`);

  const manifest = fs.readFileSync(
    path.join(wrapDir, "wrap.info")
  );
  const wasmModule = fs.readFileSync(
    path.join(wrapDir, "wrap.wasm")
  );

  console.log("Creating WrapPackage from raw wrap.info & wrap.wasm bytes");

  const wrapPackage = WasmPackage.from(
    manifest,
    wasmModule
  );

  console.log("Adding WrapPackage to ClientConfig");

  const config = new PolywrapClientConfigBuilder()
    .setPackage("embed/foo", wrapPackage)
    .build();

  const client = new PolywrapClient(config);

  console.log("Invoking WrapPackage");

  const result = await client.invoke({
    uri: "embed/foo",
    method,
    args
  });

  if (result.ok) {
    console.log("Success!");
  }
}

File: clients/py/src/features/config_embed_wrap_package.py
from pathlib import Path
from typing import Any

from polywrap import PolywrapClient, PolywrapClientConfigBuilder, Uri, WasmPackage
from pydantic import BaseModel, validator

from validators import validate_root_directory


class TestCaseInput(BaseModel):
    directory: Path
    method: str
    args: Any

    @validator("directory")
    def valid_root_directory(cls, v: Any):
        return validate_root_directory(
            v, Path(__file__).parent.parent.parent.parent.parent
        )


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.validate(input)

    print(f"Reading wrap.info & wrap.wasm from {input['directory']}")

    with open(Path(input_obj.directory, "wrap.info"), "rb") as file:
        manifest = file.read()

    with open(Path(input_obj.directory, "wrap.wasm"), "rb") as file:
        wasm_module = file.read()

    print("Creating WrapPackage from raw wrap.info & wrap.wasm bytes")

    wrap_package = WasmPackage(NotImplemented, manifest, wasm_module)

    print("Adding WrapPackage to ClientConfig")

    config = (
        PolywrapClientConfigBuilder()
        .set_package(Uri("embed", "str"), wrap_package)
        .build()
    )
    client = PolywrapClient(config)

    print("Invoking WrapPackage")

    try:
        client.invoke(
            uri=Uri.from_str("embed/str"), method=input_obj.method, args=input_obj.args
        )
        print("Success!")
    except Exception as e:
        print(e)
